core/chains/evm/txmgr/finalizer.go

```diff
@@ -0,0 +1,275 @@
+}
+
+func (f *evmFinalizer) HealthReport() map[string]error {
+	return map[string]error{f.Name(): f.Healthy()}
+}
+
+func (f *evmFinalizer) runLoop() {
+	defer f.wg.Done()
+	ctx, cancel := f.stopCh.NewCtx()
+	defer cancel()
+	for {
+		select {
+		case <-f.mb.Notify():
+			for {
+				if ctx.Err() != nil {
+					return
+				}
+				head, exists := f.mb.Retrieve()
+				if !exists {
+					break
+				}
+				if err := f.ProcessHead(ctx, head); err != nil {
+					f.lggr.Errorw("Error processing head", "err", err)
+					f.SvcErrBuffer.Append(err)
+					continue
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *evmFinalizer) DeliverLatestHead(head *evmtypes.Head) bool {
+	return f.mb.Deliver(head)
+}
+
+func (f *evmFinalizer) ProcessHead(ctx context.Context, head *evmtypes.Head) error {
+	ctx, cancel := context.WithTimeout(ctx, processHeadTimeout)
+	defer cancel()
+	_, latestFinalizedHead, err := f.headTracker.LatestAndFinalizedBlock(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve latest finalized head: %w", err)
+	}
+	return f.processFinalizedHead(ctx, latestFinalizedHead)
+}
+
+// Determines if any confirmed transactions can be marked as finalized by comparing their receipts against the latest finalized block
+func (f *evmFinalizer) processFinalizedHead(ctx context.Context, latestFinalizedHead *evmtypes.Head) error {
+	// Cannot determine finality without a finalized head for comparison
+	if latestFinalizedHead == nil || !latestFinalizedHead.IsValid() {
+		return fmt.Errorf("invalid latestFinalizedHead")
+	}
+	earliestBlockNumInChain := latestFinalizedHead.EarliestHeadInChain().BlockNumber()
+	f.lggr.Debugw("processing latest finalized head", "block num", latestFinalizedHead.BlockNumber(), "block hash", latestFinalizedHead.BlockHash(), "earliest block num in chain", earliestBlockNumInChain)
+
+	// Retrieve all confirmed transactions, loaded with attempts and receipts
+	unfinalizedTxs, err := f.txStore.FindConfirmedTxes(ctx, f.chainId)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve confirmed transactions: %w", err)
+	}
+
+	var finalizedTxs []*Tx
+	// Group by block hash transactions whose receipts cannot be validated using the cached heads
+	receiptBlockHashToTx := make(map[common.Hash][]*Tx)
+	// Find transactions with receipt block nums older than the latest finalized block num and block hashes still in chain
+	for _, tx := range unfinalizedTxs {
+		receipt := tx.GetReceipt()
+		if receipt == nil || receipt.IsZero() || receipt.IsUnmined() {
+			f.lggr.AssumptionViolationw("invalid receipt found for confirmed transaction", "tx", tx, "receipt", receipt)
+			continue
+		}
+		// Receipt newer than latest finalized head block num
+		if receipt.GetBlockNumber().Cmp(big.NewInt(latestFinalizedHead.BlockNumber())) > 0 {
+			continue
+		}
+		// Receipt block num older than earliest head in chain. Validate hash using RPC call later
+		if receipt.GetBlockNumber().Int64() < earliestBlockNumInChain {
```

Comment:
Is this to protect against local finality violations? What's the reasoning behind this? I would expect to check for every receipt that is equal or lower to the finalized block.